[WIP] Get entity at instant

TheodorRene opened the pull request:
WIP at the moment. Included two implementations, one naive, and one smarter. Will maybe try and see which is fastest or if its neglible. 

Welcome to the mother of all PRs

- [x] Write tests for getEntity[Edge]
- [ ] Write implementation for getEdge(srcId,dstId)

TheodorRene committed:
Add getEntity to TemporalGraph

TheodorRene committed:
Move documentation to the correct place

TheodorRene committed:
[WIP] Add initial implementation of getEntity for snapshotDelta

eivindkopperud commented on data-generation/src/main/scala/thesis/SnapshotDelta.scala:
```diff
@@ -68,8 +69,43 @@ class SnapshotDelta(val graphs: MutableList[Snapshot],
+  // Possibly smarter way
+  // No tests written yet
+  def getVertexSmart(vertex: VERTEX, instant: Instant): Option[(Entity, Attributes)] = {
+    val Snapshot(graph, graphInstant) = getClosestGraph(graphs, instant)
```
Helt vill unpacking! 🧠

eivindkopperud reviewed (commented):
Synes `getVertexSmart` virker veldig lovende! Synes det var ganske tydelig hva man gjør der også, bra jobba!

TheodorRene committed:
Refactor getMostRecentMaterializedGraph
It does not make any sense to return a Graph if it is after the given instant. We now make that obvious to the compiler as well

TheodorRene committed:
Add mutable prefix to mutableList
Intellij tells us we have to do it

TheodorRene committed:
Add getVertex

eivindkopperud commented on data-generation/src/main/scala/utils/LogUtils.scala:
```diff
@@ -24,10 +24,22 @@ object LogUtils {
    * Order all logs as key-value pairs, with edge ids as the key and a list of corresponding logs as the value.
    *
-   * @param logs
+   * @param logs logs
```
Bare et notat: jeg synes det er irriterende at det kommer warnings her hvis man ikke har noe her. Dersom parameteret er selvforklarende så er det jo bedre å ikke skrive docs for det tenker jeg 🤷‍♂️

TheodorRene replied:
Enig, den klager vel kanskje ikke hvis man bare fjerner den?

eivindkopperud replied:
not sure! det kan være, men da får man dårligere intellisense kanskje? idk merkelig opplegg

eivindkopperud commented on data-generation/src/test/scala/SnapshotDeltaSpec.scala:
```diff
@@ -212,4 +212,21 @@ class SnapshotDeltaSpec extends AnyFlatSpec with SparkTestWrapper {
 
   }
+
+  "getVertex" should "retrieve the correct entity at the right timestamp" in {
```
Vil du mekke en tilsvarende for getEdge også?

eivindkopperud commented on data-generation/src/main/scala/thesis/SnapshotDelta.scala:
```diff
@@ -53,23 +69,97 @@ class SnapshotDelta(val graphs: MutableList[Snapshot],
+    }
+  }
+
+  // Thinking about moving these two functions to another file. Thoughts?
```
`SnapshotQueryUtils` typ? Det blir kanskje en del ryddigere ja. Skal vi gjøre det med andre queries også? Har noen ymse hjelpefunksjoner i Landy/Snapshot i en annen PR vet jeg

eivindkopperud commented on data-generation/src/main/scala/thesis/SnapshotDelta.scala:
```diff
@@ -53,23 +69,97 @@ class SnapshotDelta(val graphs: MutableList[Snapshot],
+
+  // Thinking about moving these two functions to another file. Thoughts?
+  def getMaterializedVertex(vertex: VERTEX, instant: Instant, snapshot: Snapshot): Option[(VERTEX, Attributes)] = {
+    val Snapshot(graph, mInstant) = snapshot
```
```suggestion
    val Snapshot(graph, materializedInstant) = snapshot
```
Siden det er tanken ja? Synes personlig dette er mer lesbart, men up to you

eivindkopperud reviewed (commented):
Nice!! Ganske tungvint spørring (av natur), men synes du har gjort en veldig god jobb på å gjøre dette forståelig! 👏

TheodorRene committed:
Refactor Entity into Vertex and Edge sharing an id field

This makes a lot of functions that can be identical for vertices and edges much easier to write

TheodorRene committed:
Refactor getVertex and getEdge into getEntity with some type magic

This saves means we can reuse a lot of code

TheodorRene commented:
Refactor av Entity burde egentlig vært en egen PR, men var så dypt inni det og det var vanskelig å skille ut

eivindkopperud commented on data-generation/src/main/scala/thesis/SnapshotDelta.scala:
```diff
@@ -91,54 +91,56 @@ class SnapshotDelta(val graphs: mutable.MutableList[Snapshot],
   /** get Vertex at a certain point in time
```
```suggestion
  /** Get the given entity at the given point in time
```

eivindkopperud commented on data-generation/src/main/scala/thesis/SnapshotDelta.scala:
```diff
@@ -91,54 +91,56 @@ class SnapshotDelta(val graphs: mutable.MutableList[Snapshot],
-    val possibleVertexAttributes = graph.vertices.lookup(vertex.objId).headOption
-    val relevantLogs = LogUtils.filterVertexLogsById(getLogsInInterval(logs, Interval(mInstant, instant)), vertex)
-    val possibleLog = getSquashedActionsByVertexId(relevantLogs).lookup(vertex.objId).headOption
+    val possibleVertexAttributes = graph.vertices.lookup(entity.id).headOption
```
```suggestion
    val possibleEntityAttributes = graph.vertices.lookup(entity.id).headOption
```

eivindkopperud commented on data-generation/src/main/scala/thesis/SnapshotDelta.scala:
```diff
@@ -91,54 +91,56 @@ class SnapshotDelta(val graphs: mutable.MutableList[Snapshot],
+    val relevantLogs = LogUtils.filterEntityLogsById(getLogsInInterval(logs, Interval(mInstant, instant)), entity)
+    val possibleLog = getSquashedActionsByVertexId(relevantLogs).lookup(entity.id).headOption
     (possibleVertexAttributes, possibleLog) match {
       case (Some(vertexAttributes), Some(log)) => log.action match {
```
```suggestion
      case (Some(entityAttributes), Some(log)) => log.action match {
```

eivindkopperud commented on data-generation/src/main/scala/thesis/DataTypes.scala:
```diff
@@ -87,13 +87,13 @@ object Action {
 
-object Entity {
-  final case class VERTEX(objId: Long) extends Entity
+final case class VERTEX(override val id: Long) extends Entity
```
```suggestion
final case class VERTEX(id: Long) extends Entity
```
Does this work?

TheodorRene replied:
Nei. https://stackoverflow.com/questions/12705309/scala-case-class-inheritance Men jeg kunne kanskje valgt det første svaret. Tar det eventuelt på en annen PR

TheodorRene committed:
Merge branch 'main' into feat/entitiyAtPoint

The merge was done because there were so many merge conflicts and it
would take me ages to fix everyone of them

TheodorRene committed:
Add missing import

eivindkopperud commented:
> - [ ] Write implementation for getEdge(srcId,dstId)

Kan dette komme i ny pr?<3

TheodorRene committed:
Fix badly implemented getEntities and add test for getEntity[Edge]

TheodorRene committed:
Vertex -> Entities because of more generic function
